test/integration/buckets-client-condreq.test.js

```diff
@@ -118,66 +154,375 @@ test('buckets client conditional requests', testOpts, function (suite) {
+
+            t.ok(err);
+            t.ok(res);
+
+            t.equal(err.code, 'PreconditionFailed');
+
+            headAndAssert(t, function (res) {
+                t.equal(res.headers['m-foo'], 'bar');
+                t.end();
+            });
+        });
+    });
+
+    test('CreateBucketObject: if-none-match (bad)', function (t) {
+        var inStream = fs.createReadStream(SMALL_FILE_PATH);
+        var reqOpts = {
+            headers: {
+                'm-bar': 'wut',
+                'if-none-match': etag
+            }
+        };
+        client.createBucketObject(inStream, BUCKET_NAME, OBJECT_NAME, reqOpts,
+                                  function (err, res) {
+
+            t.ok(err);
+            t.ok(res);
+
+            t.equal(err.code, 'PreconditionFailed');
+
+            headAndAssert(t, function (res) {
+                t.equal(res.headers['m-foo'], 'bar');
+                t.end();
+            });
+        });
+    });
+
+    test('CreateBucketObject: if-match (good)', function (t) {
+        var inStream = fs.createReadStream(SMALL_FILE_PATH);
+        var reqOpts = {
+            headers: {
+                'if-match': etag,
+                'm-bar': 'wut',
+                'm-foo': 'bar'
+            }
+        };
+        client.createBucketObject(inStream, BUCKET_NAME, OBJECT_NAME, reqOpts,
+                                  function (err, res) {
+            t.ifError(err);
+            t.ok(res);
+
+            headAndAssert(t, function (res) {
+                t.equal(res.headers['m-foo'], 'bar');
+                t.equal(res.headers['m-bar'], 'wut');
+                t.ok(etag !== res.headers['etag']);
+                etag = res.headers['etag'];
+                mtime = new Date(res.headers['last-modified']);
+                t.ok(!isNaN(mtime.getTime()));
+                t.end();
+            });
+        });
+    });
+
+    /*
+     * UPDATE
+     */
+
+    test('UpdateBucketObject: if-match (good)', function (t) {
+        client.putBucketObjectMetadata(BUCKET_NAME, OBJECT_NAME, {
+            headers: {
+                'm-foo': 'wut',
+                'if-match': etag
+            } }, function (err, res) {
+
+            t.ifError(err);
+            t.ok(res);
+
+            headAndAssert(t, function (res) {
+                t.equal(res.headers['m-foo'], 'wut');
+                t.ok(!res.headers['m-bar']);
+                mtime = new Date(res.headers['last-modified']);
+                t.ok(!isNaN(mtime.getTime()));
+                t.end();
+            });
+        });
+    });
+
+    test('UpdateBucketObject: if-match (bad)', function (t) {
+        client.putBucketObjectMetadata(BUCKET_NAME, OBJECT_NAME, {
+            headers: {
+                'm-foo': 'test',
+                'if-match': libuuid.v4()
+            } }, function (err, res) {
+
+            t.ok(err);
+            t.ok(res);
+            /*
+             * XXX Why no `res.statusCode` from client, like `get` does.
```

Comment:
This is really odd. Wonder why it would be different in this case?

Reply:
Oh man, sorry about this one. I think my brain is a bit frazzled. I just spent way too long looking at why node-manta wasn't parsing a response body from a HEAD request...

So the difference here is that node-manta's buckets client does a lot of the response wrangling itself, but for GETs it leans on the existing directory client for some niceties around resumability.

The difference is that the buckets error path doesn't return `res` on error and instead populates `err` with all the information. Directories is different where `res` is included as well it seems, so this is why GET in the buckets client also includes `res`.

The latest commit should account for this in the test suite. I think there is still some work to do to make this consistent across all client requests, but it's not a million miles off and I expect explains why the buckets work is still on its own branch.